File: src/cs/Ssh/Algorithms/PublicKeyAlgorithm.cs
// Copyright (c) Microsoft. All rights reserved.

using System;
using System.Text;
using Microsoft.DevTunnels.Ssh.IO;

namespace Microsoft.DevTunnels.Ssh.Algorithms;

public abstract class PublicKeyAlgorithm : SshAlgorithm
{
	protected PublicKeyAlgorithm(
		string name,
		string keyAlgorithmName,
		string hashAlgorithmName)
		: base(name)
	{
		this.KeyAlgorithmName = keyAlgorithmName;
		this.HashAlgorithmName = hashAlgorithmName;
	}

	public string KeyAlgorithmName { get; set; }
	public string HashAlgorithmName { get; }

	public abstract IKeyPair CreateKeyPair();
	public abstract IKeyPair GenerateKeyPair(int? keySizeInBits = null);
	public abstract ISigner CreateSigner(IKeyPair keyPair);
	public abstract IVerifier CreateVerifier(IKeyPair keyPair);

	public Buffer ReadSignatureData(Buffer signatureData)
	{
		var reader = new SshDataReader(signatureData);
		string algorithmName = reader.ReadString(Encoding.ASCII);
		if (algorithmName != Name)
		{
			throw new ArgumentException(
				"Mismatched public key algorithm: " +
				$"got '{algorithmName}', expected '{Name}'.");
		}

		var signature = reader.ReadBinary();
		return signature;
	}

	public Buffer CreateSignatureData(Buffer signature)
	{
		var writer = new SshDataWriter();
		writer.Write(Name, Encoding.ASCII);
		writer.WriteBinary(signature);
		return writer.ToBuffer();
	}
}

File: src/ts/ssh/algorithms/publicKeyAlgorithm.ts
//
//  Copyright (c) Microsoft Corporation. All rights reserved.
//

import { Buffer } from 'buffer';
import { SshDataWriter, SshDataReader } from '../io/sshData';
import { SshAlgorithm } from './sshAlgorithm';
import { Signer, Verifier } from './hmacAlgorithm';
import { Disposable, CancellationToken } from 'vscode-jsonrpc';
import { BigInt } from '../io/bigInt';

export abstract class PublicKeyAlgorithm implements SshAlgorithm {
	protected constructor(
		public readonly name: string,
		public readonly keyAlgorithmName: string,
		public readonly hashAlgorithmName: string,
	) {}

	public abstract createKeyPair(): KeyPair;
	public abstract generateKeyPair(keySizeInBits?: number): Promise<KeyPair>;
	public abstract createSigner(keyPair: KeyPair): Signer;
	public abstract createVerifier(keyPair: KeyPair): Verifier;

	public readSignatureData(signatureData: Buffer): Buffer {
		const reader = new SshDataReader(signatureData);
		const algorithmName = reader.readString('ascii');
		if (algorithmName !== this.name) {
			throw new Error(
				'Mismatched public key algorithm: ' +
					`got '${algorithmName}', expected '${this.name}'.`,
			);
		}

		const signature = reader.readBinary();
		return signature;
	}

	public createSignatureData(signature: Buffer): Buffer {
		const writer = new SshDataWriter(Buffer.alloc(this.name.length + signature.length + 20));
		writer.writeString(this.name, 'ascii');
		writer.writeBinary(signature);
		return writer.toBuffer();
	}
}

export interface KeyPair extends Disposable {
	readonly keyAlgorithmName: string;
	readonly hasPublicKey: boolean;
	readonly hasPrivateKey: boolean;
	comment: string | null;
	setPublicKeyBytes(keyBytes: Buffer): Promise<void>;
	getPublicKeyBytes(algorithmName?: string): Promise<Buffer | null>;
	generate(): Promise<void>;
	importParameters(parameters: KeyPairParameters): Promise<void>;
	exportParameters(): Promise<KeyPairParameters>;
}

// eslint-disable-next-line @typescript-eslint/no-empty-interface
interface KeyPairParameters {}

export interface RsaParameters extends KeyPairParameters {
	modulus: BigInt;
	exponent: BigInt;
	d?: BigInt;
	p?: BigInt;
	q?: BigInt;
	dp?: BigInt;
	dq?: BigInt;
	qi?: BigInt;
}

export interface ECParameters extends KeyPairParameters {
	curve: { name?: string; oid?: string };
	x: BigInt;
	y: BigInt;
	d?: BigInt;
}

/**
 * Given a public key, provides the corresponding private key.
 */
export type PrivateKeyProvider = (
	publicKey: KeyPair,
	cancellation: CancellationToken,
) => Promise<KeyPair | null>;
